mongoengine/base/fields.py

```diff
@@ -275,11 +275,16 @@ def __get__(self, instance, owner):
 
         _dereference = _import_class('DeReference')()
 
-        if instance._initialised and dereference and instance._data.get(self.name):
+        if (instance._initialised and
+                dereference and
+                instance._data.get(self.name) and
+                not getattr(instance._data[self.name], '_dereferenced', False)):
```

Comment:
its minor but I think it would slightly improve readability if you set a variable outside the `if`
like
```
already_dereferenced = getattr(instance._data[self.name], '_dereferenced', False)
if (instance._initialised
        and dereference and
        instance._data.get(self.name) and
        not already_dereferenced):
        ...
```

Reply:
@bagerard I agree that would be better for readability however, we need to make the `instance._data.get(self.name)` check before we use `instance._data[self.name]` in the final check, so I've kept it as-is for now.